system/metrics/testing/metrictest.cpp

```diff
@@ -79,53 +79,63 @@ const char *testLogSinkConfigYml = R"!!(component:
+    try
+    {
+        InitModuleObjects();
 
-    //
-    // Retrieve the global and component metrics config
-    Owned<IPropertyTree> pMetricsTree = pSettings->getPropTree("component/metrics");
+        //
+        // Simulate retrieving the component and global config
+        Owned<IPropertyTree> pSettings = createPTreeFromYAMLString(testLogSinkConfigYml, ipt_none, ptr_ignoreWhiteSpace, nullptr);
 
-    //
-    // Allow override of output file for the file sink
-    if (argc > 1)
-    {
-        auto pSinkTree = pMetricsTree->getPropTree("component/metrics/sinks[1]/settings");
-        pSinkTree->setProp("@filename", argv[1]);
-    }
+        //
+        // Retrieve the global and component metrics config
+        Owned<IPropertyTree> pMetricsTree = pSettings->getPropTree("component/metrics");
+
+        //
+        // Allow override of output file for the file sink
+        if (argc > 1)
+        {
+            auto pSinkTree = pMetricsTree->getPropTree("component/metrics/sinks[1]/settings");
+            pSinkTree->setProp("@filename", argv[1]);
+        }
 
 
-    //
-    // Get singleton
-    MetricsReporter &myReporter = queryMetricsReporter();
+        //
+        // Get singleton
+        MetricsReporter &myReporter = queryMetricsReporter();
 
-    //
-    // Init reporter with config
-    myReporter.init(pMetricsTree);
+        //
+        // Init reporter with config
+        myReporter.init(pMetricsTree);
 
-    //
-    // Now create the metrics and add them to the reporter
-    pEventCountMetric = std::make_shared<CounterMetric>("requests", "The number of requests");
-    myReporter.addMetric(pEventCountMetric);
+        //
+        // Now create the metrics and add them to the reporter
+        pEventCountMetric = std::make_shared<CounterMetric>("requests", "The number of requests");
+        myReporter.addMetric(pEventCountMetric);
 
-    pQueueSizeMetric = createMetricAndAddToReporter<GaugeMetric>("queuesize", "request queue size");
+        pQueueSizeMetric = createMetricAndAddToReporter<GaugeMetric>("queuesize", "request queue size");
 
-    myReporter.startCollecting();
+        myReporter.startCollecting();
 
-    //
-    // Starts some threads, each updating metrics
-    std::thread first (processThread, 20, 2, true, "requests_dynamic", 4, 10);
-    std::thread second (processThread, 15, 3, false, "", 0, 0);
+        //
+        // Starts some threads, each updating metrics
+        std::thread first (processThread, 20, 2, true, "requests_dynamic", 4, 10);
+        std::thread second (processThread, 15, 3, false, "", 0, 0);
 
-    first.join();
-    second.join();
+        first.join();
+        second.join();
+
+        printf("Stopping the collection...");
+        myReporter.stopCollecting();
+        printf("Stopped. Test complete\n");
+    }
+
+    catch (...)
+    {
+        printf("Exception detected, test stopped");
+    }
```

Comment:
There might be some missing cleanup code, look into these:
releaseAtoms();
ExitModuleObjects();

Also, is there a chance MetricsReporter needs to shutdown?

Reply:
@rpastrana I don't create any atoms. Also, the InitModuleObjects registers a call to ExitModuleObjects automatically

#define InitModuleObjects() { _InitModuleObjects(); atexit(&ExitModuleObjects); }

MetricsReporter does all it's shutdown during deconstruction.